homeassistant/components/nissan_leaf.py

```diff
@@ -0,0 +1,489 @@
+                self.data[DATA_RANGE_AC] = (
+                    battery_response.cruising_range_ac_on_km
+                )
+                self.data[DATA_RANGE_AC_OFF] = (
+                    battery_response.cruising_range_ac_off_km
+                )
+                self.signal_components()
+                self.last_battery_response = utcnow()
+
+        if server_response is not None:
+            _LOGGER.debug("Server Response: ")
+            _LOGGER.debug(server_response.__dict__)
+
+            if server_response.answer['status'] == 200:
+                if server_response.state_of_charge is not None:
+                    self.data[DATA_BATTERY] = int(
+                        server_response.state_of_charge
+                    )
+                else:
+                    self.data[DATA_BATTERY] = server_response.battery_percent
+
+                self.data[DATA_RANGE_AC] = (
+                    server_response.cruising_range_ac_on_km
+                )
+                self.data[DATA_RANGE_AC_OFF] = (
+                    server_response.cruising_range_ac_off_km
+                )
+                self.signal_components()
+                self.last_battery_response = utcnow()
+
+        # Climate response only updated if battery data updated first.
+        if (battery_response is not None) or (server_response is not None):
+
+            climate_response = await self.async_get_climate()
+            if climate_response is not None:
+                _LOGGER.debug("Got climate data for Leaf.")
+                _LOGGER.debug(climate_response.__dict__)
+                self.data[DATA_CLIMATE] = climate_response.is_hvac_running
+
+        if self.nissan_connect:
+            try:
+                location_response = await self.async_get_location()
+
+                if location_response is None:
+                    _LOGGER.debug("Empty Location Response Received")
+                    self.data[DATA_LOCATION] = None
+                else:
+                    _LOGGER.debug("Got location data for Leaf")
+                    self.data[DATA_LOCATION] = location_response
+
+                    _LOGGER.debug("Location Response: ")
+                    _LOGGER.debug(location_response.__dict__)
+            except CarwingsError:
+                _LOGGER.error("Error fetching location info")
+
+        self.request_in_progress = False
+        self.signal_components()
+
+    def signal_components(self):
+        """Signal components to refresh."""
+        async_dispatcher_send(self.hass, SIGNAL_UPDATE_LEAF)
+
+    async def async_get_battery(self):
+        """Request battery update from Nissan servers."""
+        # First, check nissan servers for the latest data
+        start_server_info = await self.hass.async_add_job(
+            self.leaf.get_latest_battery_status
+        )
+
+        # Store the date from the nissan servers
+        start_date = start_server_info.answer[
```

Comment:
I would probably add a try ... catch around this (just incase the server responds something different in the future)

Reply:
I've put a try/catch around the whole of async_get_battery.